### PR TITLE
Add variable `grafana_ingress_host` to expose Grafana

### PR DESCRIPTION
**Pull request by surajssd**
Fixes: https://github.com/kinvolk/lokomotive/issues/456

### CODE REVIEW

File: pkg/components/prometheus-operator/template.go [hunk 1]
```diff
@@ -55,6 +57,21 @@ grafana:
       searchNamespace: ALL
   rbac:
     pspUseAppArmor: false
+
+  ingress:
+    enabled: {{ .ExposeGrafana }}
```
Comment (invidian) [inline pkg/components/prometheus-operator/template.go:62]: Maybe just `{{printf "%t" .GrafanaIngressHost != ""}}` instead of separate filed?

File: docs/configuration-reference/components/prometheus-operator.md [hunk 2]
```diff
@@ -29,8 +29,10 @@ Prometheus Operator component configuration example:
 
 ```tf
 component "prometheus-operator" {
-  namespace              = "monitoring"
+  namespace = "monitoring"
+
   grafana_admin_password = "foobar"
+  grafana_ingress_host = "grafana.mydomain.net"
```
Comment (invidian) [inline docs/configuration-reference/components/prometheus-operator.md:35]:
```suggestion
  grafana_admin_password = "foobar"
  grafana_ingress_host   = "grafana.mydomain.net"
```

File: docs/configuration-reference/components/prometheus-operator.md [hunk 3]
```diff
@@ -86,6 +88,7 @@ Example:
 |--------	|--------------|:-------:|:--------:|
 | `namespace` | Namespace to deploy the Prometheus Operator. | - | true |
 | `grafana_admin_password` | Password for `admin` user in Grafana.  | - | true |
+| `grafana_ingress_host` | Ingress URL Host to expose Grafana over internet. **NOTE:** User will have to create a DNS entry for this against Packet EIP manually if running on Packet.  | - | false |
```
Comment (johananl) [inline docs/configuration-reference/components/prometheus-operator.md:91]: >**NOTE:** User will have to create a DNS entry for this against Packet EIP manually if running on Packet.

It feels weird to me to include a note in the knob's description. Is there a better place to put this info?

I suggest the following alternative phrasing: "When running on Packet, a DNS entry pointing at the ingress controller needs to be created." I think it's better because of the following:

- "User will have to..." is IMO the wrong tone for using in documentation. We either use the imperative tense "Do this and that", or we use the passive tense "This and that needs to be done".
- "Packet EIP" is ambiguous and assumes context - we assume the user knows what an EIP is and that we are talking specifically about the EIP used for Contour. We should be more concrete.
- There is a missing article ("the") before "Packet WIP".
- The condition ("if running on Packet") should come before the instruction to avoid forcing the user to read a block of text only to discover it doesn't apply to their situation when reaching its end ("I'm running on AWS").

File: docs/configuration-reference/components/prometheus-operator.md [hunk 4]
```diff
@@ -29,8 +29,10 @@ Prometheus Operator component configuration example:
 
 ```tf
 component "prometheus-operator" {
-  namespace              = "monitoring"
+  namespace = "monitoring"
+
   grafana_admin_password = "foobar"
+  grafana_ingress_host   = "grafana.mydomain.net"
```
[the remark this answers is not in the document]
Comment (surajssd) [reply docs/configuration-reference/components/prometheus-operator.md:35]: > Can't we use a convention such as grafana.<cluster_name>.<zone> rather than add a new knob? At the very least we could use the convention by default and allow an explicit override if this needs to be user configurable.

1. Prometheus Operator has no knowledge of the cluster it is being deployed upon. So there is no way it can enforce upon above information without having to ask user again for `cluster_name` and `zone`. This is because there is no wiring between the cluster config and the components config behind the scene. Making that work will be bigger a refactor work.

2. I am not sure if a component should be enforcing such a convention, users are totally free to select whatever they want to. But if you think it makes sense to be opionated about it then above problem still holds true.

File: docs/configuration-reference/components/prometheus-operator.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (surajssd) [reply docs/configuration-reference/components/prometheus-operator.md:91]: If we do that then we will need one more knob which will decide if we want to expose the grafana via ingress. Right now the way config is created. If user does not provide anything then that means user does not want to expose the grafana and vice-a-versa.

Comment (surajssd) [reply docs/configuration-reference/components/prometheus-operator.md:91]: > If we create a wildcard record pointing at Contour on AWS, why don't we do it on all platforms?

This will solve the problem for Packet yes. And it will make things easier for the user on Packet.

Comment (surajssd) [reply docs/configuration-reference/components/prometheus-operator.md:91]: Created an issue for this https://github.com/kinvolk/lokomotive/issues/470

Comment (surajssd) [reply docs/configuration-reference/components/prometheus-operator.md]: Made changes as you have suggested.

File: docs/configuration-reference/components/prometheus-operator.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (johananl) [reply docs/configuration-reference/components/prometheus-operator.md:35]: Yes, I'm aware we don't have this "wiring". I guess my question was "is this feasible", but I understand it's complicated and out of scope.

>I am not sure if a component should be enforcing such a convention, users are totally free to select whatever they want to. But if you think it makes sense to be opionated about it then above problem still holds true.

My point was that Prometheus/Grafana are kind of "cluster-related" and therefore it _could_ make sense to have a convention such as the one suggested above for their URLs. I even recall something about `kubectl cluster-info` showing a Grafana URL. But my main motivation for suggesting this was to reduce the number of required knobs the user needs to worry about. Adding a knob to the config is "expensive" in terms of UX.

File: docs/configuration-reference/components/prometheus-operator.md [hunk 5]
```diff
@@ -86,6 +88,7 @@ Example:
 |--------	|--------------|:-------:|:--------:|
 | `namespace` | Namespace to deploy the Prometheus Operator. | - | true |
 | `grafana_admin_password` | Password for `admin` user in Grafana.  | - | true |
+| `grafana_ingress_host` | Ingress URL Host to expose Grafana over internet. **NOTE:** When running on Packet, a DNS entry pointing at the ingress controller needs to be created.  | - | false |
```
Comment (johananl) [inline docs/configuration-reference/components/prometheus-operator.md:91]: - Could you please fix the capitalization?
- There's a missing article before "internet".

File: pkg/components/prometheus-operator/component.go [hunk 6]
```diff
@@ -46,9 +46,11 @@ type CoreDNS struct {
 }
 
 type component struct {
-	GrafanaAdminPassword string   `hcl:"grafana_admin_password,attr"`
-	Namespace            string   `hcl:"namespace,optional"`
-	EtcdEndpoints        []string `hcl:"etcd_endpoints,optional"`
+	GrafanaAdminPassword string `hcl:"grafana_admin_password,attr"`
+	GrafanaIngressHost   string `hcl:"grafana_ingress_host,optional"`
```
Comment (johananl) [inline pkg/components/prometheus-operator/component.go:50]: What happens when this knob isn't specified? Should this be documented in any way?

File: pkg/components/prometheus-operator/template.go [hunk 7]
```diff
@@ -46,6 +46,8 @@ alertmanager:
           resources:
             requests:
               storage: "{{.AlertManagerStorageSize}}"
+
+
```
Comment (johananl) [inline pkg/components/prometheus-operator/template.go:50]: I'd use a single blank line for consistency with the rest of this file.

File: pkg/components/prometheus-operator/component.go (hunk 6 above)
Comment (surajssd) [reply pkg/components/prometheus-operator/component.go:50]: If user does not specify anything then ingress won't be created for the user.

File: docs/configuration-reference/components/prometheus-operator.md (hunk 5 above)
Comment (surajssd) [reply docs/configuration-reference/components/prometheus-operator.md:91]: > Could you please fix the capitalization?

I see that only `Host` has unnecessary capitalisation?

Comment (johananl) [reply docs/configuration-reference/components/prometheus-operator.md]: Yup, looks good now.

File: docs/configuration-reference/components/prometheus-operator.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (johananl) [reply docs/configuration-reference/components/prometheus-operator.md]: I meant that logically a knob is either required or not required. If it's not required, it means it has a default value which is used when the knob isn't specified by the user. In this case, sounds like the default value is `""`. If that's the case, we should document the default accordingly. Am I missing something?

File: pkg/components/prometheus-operator/component.go (hunk 6 above)
Comment (johananl) [reply pkg/components/prometheus-operator/component.go:50]: Makes sense. I'll leave it up to you to decide whether we should document this fact. AFAICT this isn't self-evident.

**Comment by johananl** [comment]
```
TestAWSIngress: aws_test.go:49: got an HTTP error: Get "https://httpbin.ci1590498787-sj.test.lokomotive-k8s.net/get": dial tcp 3.121.121.100:443: connect: connection refused
    TestAWSIngress: aws_test.go:67: could not get a successful HTTP response in time
```

File: pkg/components/prometheus-operator/component.go (hunk 6 above)
Comment (surajssd) [reply pkg/components/prometheus-operator/component.go:50]: Added this line in the docs: `If left empty Ingress is not created.`

Comment (surajssd) [reply pkg/components/prometheus-operator/component.go:50]: Also added it in the code, template.

File: docs/configuration-reference/components/prometheus-operator.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (surajssd) [reply docs/configuration-reference/components/prometheus-operator.md]: Added it to the docs.

File: docs/configuration-reference/components/prometheus-operator.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (surajssd) [reply docs/configuration-reference/components/prometheus-operator.md:35]: We can also change the whole thing to look like following:

```tf
  grafana {
    admin_password = ""

    ingress {
      host = ""

      # Needed later
      ingress_class = ""
      certmanager_cluster_issuer = ""
    }
  }
```

This is future proof as well, WDYT?

[the remark this answers is not in the document]
Comment (surajssd) [reply docs/configuration-reference/components/prometheus-operator.md:35]: If the user does not specify ingress block at all then ingress is not created and if user does then `host` field is mandatory.

File: pkg/components/prometheus-operator/template.go [hunk 8]
```diff
@@ -56,6 +57,22 @@ grafana:
...
     pspUseAppArmor: false
+
+  ingress:
+    # If user does not provide any ingress host value then following will be 'false', thus disabling
+    # the ingress section.
+    enabled: {{printf "%t" (ne .GrafanaIngressHost "")}}
+    annotations:
+      kubernetes.io/ingress.class: contour
+      kubernetes.io/tls-acme: "true"
+      cert-manager.io/cluster-issuer: "letsencrypt-production"
```
Comment (iaguis) [inline pkg/components/prometheus-operator/template.go:68]: Should we parameterize this too like we did in #437?

Comment (surajssd) [reply pkg/components/prometheus-operator/template.go:68]: That's what I have asked in https://github.com/kinvolk/lokomotive/pull/468#discussion_r430918969

**Comment by surajssd** [comment]
So it was decided in a meeting that we create block for grafana like this:

```tf
component "prometheus-operator" {
...
  grafana {
    admin_password = ""

    ingress {
      host = ""

      # In future
      ingress_class = ""
      certmanager_cluster_issuer = ""
    }
  }
}
```

File: pkg/components/prometheus-operator/component.go [hunk 9]
```diff
@@ -45,10 +45,36 @@ type CoreDNS struct {
 	Selector map[string]string `hcl:"selector,optional"`
 }
 
+// Ingress is a generic object for specifying Kubernetes ingress manifest.
+type Ingress struct {
```
Comment (invidian) [inline pkg/components/prometheus-operator/component.go:49]: This should go to the separate package. Maybe just `pkg/components` for now?

File: pkg/components/prometheus-operator/component.go [hunk 10]
```diff
@@ -133,6 +160,10 @@ func (c *component) RenderManifests() (map[string]string, error) {
 		return nil, errors.Wrap(err, "load chart from assets")
 	}
 
+	if c.Grafana != nil && c.Grafana.Ingress != nil {
```
Comment (invidian) [inline pkg/components/prometheus-operator/component.go:163]: I think this should be done in `LoadConfig()`.

File: pkg/components/prometheus-operator/component.go [hunk 11]
```diff
@@ -45,10 +45,36 @@ type CoreDNS struct {
...
+	Host                     string `hcl:"host"`
+	Class                    string `hcl:"class,optional"`
+	CertManagerClusterIssuer string `hcl:"certmanager_cluster_issuer,optional"`
+}
+
+// SetDefaults sets default values for Ingress object only when user has not provided any
+// information.
+func (ing *Ingress) SetDefaults() {
+	if ing.Class == "" {
+		ing.Class = "contour"
```
Comment (invidian) [inline pkg/components/prometheus-operator/component.go:59]: Probably `"contour"` and `"letsencrypt-production"` should go to constants.

File: pkg/components/prometheus-operator/component_test.go [hunk 12]
```diff
@@ -27,16 +27,19 @@ func TestEmptyConfig(t *testing.T) {
 	emptyConfig := hcl.EmptyBody()
 	evalContext := hcl.EvalContext{}
 	diagnostics := c.LoadConfig(&emptyConfig, &evalContext)
-	if !diagnostics.HasErrors() {
-		t.Fatalf("Empty config should return error")
+
+	if diagnostics.HasErrors() {
```
Comment (invidian) [inline pkg/components/prometheus-operator/component_test.go:31]: Why is this needed? What happens if you don't specify grafana admin password? Could you elaborate?

File: ci/aks/aks-cluster.lokocfg.envsubst [hunk 13]
```diff
@@ -71,8 +67,9 @@ cluster "aks" {
 }
 
 component "prometheus-operator" {
-  grafana_admin_password = var.grafana_admin_password
-  disable_webhooks       = true
+  grafana {}
```
Comment (invidian) [inline ci/aks/aks-cluster.lokocfg.envsubst:70]: Do we need this block here, then?

**Comment by surajssd** [comment]
> Overall LGTM, but I don't understand, why do we remove Grafana admin password. @surajssd please add some explanation to it, so we can discuss it.

Grafana admin password should not be a compulsory option. If not provided the chart automatically creates one for you, just like Rook Ceph.

File: pkg/components/prometheus-operator/component.go (hunk 9 above)
Comment (surajssd) [reply pkg/components/prometheus-operator/component.go:49]: Moved to the existing common types dir.

File: pkg/components/prometheus-operator/component_test.go (hunk 12 above)
Comment (surajssd) [reply pkg/components/prometheus-operator/component_test.go:31]: The password is automatically generated by the chart, it should be an optional field.

**Comment by invidian** [comment]
>Grafana admin password should not be a compulsory option. If not provided the chart automatically creates one for you, just like Rook Ceph.

Hm, okay. I think we should have a bit more documentation how to obtain this password then, so it is not so cryptic to the users.

**Comment by surajssd** [comment]
> Hm, okay. I think we should have a bit more documentation how to obtain this password then, so it is not so cryptic to the users.

That I plan to handle as a part of https://github.com/kinvolk/lokomotive/pull/480

**Comment by iaguis** [comment]
I wonder if it makes sense to allow exposing grafana and allow having an insecure default password so easily. So if a user does

```
  grafana {
    ingress {
      host                       = "grafana.mydomain.net"
      class                      = "contour"
      certmanager_cluster_issuer = "letsencrypt-production"
    }
  }
```

and this exposes grafana with admin:admin password, this is surprising and insecure. Should we do something about it?

**Comment by surajssd** [comment]
> I wonder if it makes sense to allow exposing grafana and allow having an insecure default password so easily.

@iaguis there is a default password but it is not `admin`. The password if not specified is generated by helm automatically and stored in a secret, as mentioned in the docs.

File: pkg/components/util/types.go [hunk 14]
```diff
@@ -78,3 +78,27 @@ func (n *NodeSelector) Render() (string, error) {
 
 	return string(b), nil
 }
+
```
Comment (invidian) [inline pkg/components/util/types.go:81]: We should not add more things to `util` package. Just `pkg/components` will be fine. See #290 for more details.

File: pkg/components/prometheus-operator/component.go [hunk 15]
```diff
@@ -115,16 +122,26 @@ func newComponent() *component {
 				"tier":    "control-plane",
 			},
 		},
+		Grafana: &Grafana{},
```
Comment (invidian) [inline pkg/components/prometheus-operator/component.go:125]: Do we need it here?

**Comment by iaguis** [comment]
> @iaguis there is a default password but it is not `admin`. The password if not specified is generated by helm automatically and stored in a secret, as mentioned in the docs.

Ah, all good then :+1:

File: pkg/components/prometheus-operator/component.go (hunk 15 above)
Comment (surajssd) [reply pkg/components/prometheus-operator/component.go:126]: Yes, we do need it. Otherwise you will see some nil pointer exceptions.

File: pkg/components/util/types.go (hunk 14 above)
Comment (surajssd) [reply pkg/components/util/types.go]: Moved to `pkg/types`.

Comment (invidian) [reply pkg/components/util/types.go]: Fair enough :smile:

File: pkg/components/prometheus-operator/component.go (hunk 15 above)
Comment (invidian) [reply pkg/components/prometheus-operator/component.go:126]: I think those can be avoided by checking them then? It doesn't make much sense to me to put it here.

Comment (invidian) [reply pkg/components/prometheus-operator/component.go:126]: Tests are also passing if I remove it.

Comment (invidian) [reply pkg/components/prometheus-operator/component.go:126]: Ah, I'm able to trigger it when I modified the tests, hmm.

Comment (invidian) [reply pkg/components/prometheus-operator/component.go:126]: Maybe we can make Grafana block not a pointer?

Comment (surajssd) [reply pkg/components/prometheus-operator/component.go:126]: What's wrong with it? What do you think would be potential implications if we don't make it a pointer?

**Comment by invidian** [comment]
How about something like this?

```diff
diff --git a/pkg/components/prometheus-operator/component.go b/pkg/components/prometheus-operator/component.go
index b3791da3..137b997f 100644
--- a/pkg/components/prometheus-operator/component.go
+++ b/pkg/components/prometheus-operator/component.go
@@ -123,7 +123,6 @@ func newComponent() *component {
                                "tier":    "control-plane",
                        },
                },
-               Grafana: &Grafana{},
        }
 }

diff --git a/pkg/components/prometheus-operator/component_test.go b/pkg/components/prometheus-operator/component_test.go
index 8e2608fb..b0b6cd39 100644
--- a/pkg/components/prometheus-operator/component_test.go
+++ b/pkg/components/prometheus-operator/component_test.go
@@ -26,18 +26,25 @@ func TestEmptyConfig(t *testing.T) {
        c := newComponent()
        emptyConfig := hcl.EmptyBody()
        evalContext := hcl.EvalContext{}
-       diagnostics := c.LoadConfig(&emptyConfig, &evalContext)

-       if diagnostics.HasErrors() {
+       if d := c.LoadConfig(&emptyConfig, &evalContext); d.HasErrors() {
                t.Fatalf("Empty config should not return errors")
        }
+
+       m, err := c.RenderManifests()
+       if err != nil {
+               t.Fatalf("Rendering manifests with valid config should succeed, got: %s", err)
+       }
+       if len(m) <= 0 {
+               t.Fatalf("Rendered manifests shouldn't be empty")
+       }
 }

 func TestRenderManifest(t *testing.T) {
        configHCL := `
 component "prometheus-operator" {
   grafana {
-       admin_password = "foobar"
+    admin_password = "foobar"
   }
   namespace = "monitoring"
 }
diff --git a/pkg/components/prometheus-operator/template.go b/pkg/components/prometheus-operator/template.go
index 3a96fe0f..7089874c 100644
--- a/pkg/components/prometheus-operator/template.go
+++ b/pkg/components/prometheus-operator/template.go
@@ -48,7 +48,6 @@ alertmanager:
               storage: "{{.AlertManagerStorageSize}}"

 grafana:
-  adminPassword: {{.Grafana.AdminPassword}}
   plugins: "grafana-piechart-panel"
   testFramework:
     enabled: false
@@ -57,6 +56,8 @@ grafana:
       searchNamespace: ALL
   rbac:
     pspUseAppArmor: false
+  {{- if .Grafana}}
+  adminPassword: {{.Grafana.AdminPassword}}
   {{ if .Grafana.Ingress }}
   ingress:
     enabled: true
@@ -71,6 +72,7 @@ grafana:
       - {{ .Grafana.Ingress.Host }}
       secretName: {{ .Grafana.Ingress.Host }}-tls
   {{ end }}
+  {{- end}}

 kubeEtcd:
   enabled: {{.Monitor.Etcd}}
```

File: pkg/components/prometheus-operator/component.go (hunk 15 above)
Comment (invidian) [reply pkg/components/prometheus-operator/component.go:126]: I'm not sure about the possible implications, but if it's a pointer, it means we should handle the cases when it's `nil`. Also, if defined here, if provides no "default" values, as we usually use `newComponent` for, so I'd remove it.

Please have a look at the patch I posted. It adds the handling and test for it.

**Comment by surajssd** [comment]
@invidian PTAL, added table driven tests. Also taken care of the test that fails.

File: pkg/components/prometheus-operator/component_test.go [hunk 16]
```diff
@@ -17,46 +17,77 @@ package prometheus
...
+  grafana {
+    ingress {}
+  }
+}`,
+			wantErr: true,
+		},
+	}
 
-	component := newComponent()
+	for _, tc := range tests {
```
Comment (invidian) [inline pkg/components/prometheus-operator/component_test.go:67]: please use `t.Run` when doing table driven tests.

File: pkg/components/prometheus-operator/component_test.go [hunk 17]
```diff
@@ -17,46 +17,77 @@ package prometheus
...
-	if err != nil {
-		t.Fatalf("Rendering manifests with valid config should succeed, got: %s", err)
-	}
-	if len(m) <= 0 {
-		t.Fatalf("Rendered manifests shouldn't be empty")
+		if tc.wantErr && !d.HasErrors() {
+			t.Errorf("%s - Wrong config should have returned an error", tc.desc)
+		}
+
+		m, err := c.RenderManifests()
```
Comment (invidian) [inline pkg/components/prometheus-operator/component_test.go:84]: If I want error, meaning loading the HCL failed, this should not be executed, or?

Comment (surajssd) [reply pkg/components/prometheus-operator/component_test.go:84]: By want error it means that the test is expecting an error.

Comment (invidian) [reply pkg/components/prometheus-operator/component_test.go:84]: yes, sorry. I mean, if we fail to load the HCL, the component will be in the possibly undefined state, so we should not try to render it then.

Comment (surajssd) [reply pkg/components/prometheus-operator/component_test.go]: Yes changed that now it is using t.Run so now I can return using `t.Fatal`.

File: pkg/components/prometheus-operator/component_test.go (hunk 16 above)
Comment (surajssd) [reply pkg/components/prometheus-operator/component_test.go:67]: done.

File: pkg/components/prometheus-operator/component.go [hunk 18]
```diff
@@ -115,16 +123,26 @@ func newComponent() *component {
...
```
Comment (invidian) [inline pkg/components/prometheus-operator/component.go:126]: With the template changed, we no longer need this line.

**Comment by surajssd** [comment]
> LGTM. One last thing which can be done independently of this patch, I guess we should test if exposing ingress works as expected via e2e test?

Testing on packet can be tricky because:

- Someone has to expose the EIP on Route53.
- How do we use unique EIP per CI run? And make corresponding entries?

**Comment by invidian** [comment]
The testing can be done on any platform. But should be done at at least one.
